lib/builtins/completion.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+require 'builtin'
+require 'forwards'
+require 'builtins/helpers/enumerator'
+require 'options'
+
+module Builtins
+	class Completion < Builtin
+		BASH = "bash"
+		ZSH = "zsh"
+		USAGE = "Usage: ops completion 'bash / zsh'"
+
+		class << self
+			def description
+				"displays completion configuration for the flavor of shell provided"
+			end
+		end
+
+		def run
+			if args.none? || ![BASH, ZSH].include?(args[0])
+				Output.error(USAGE)
+				false
+			elsif args[0] == BASH
+				Output.print(bash_completion)
+				true
+			elsif args[0] == ZSH
+				Output.print(zsh_completion)
+				true
+			end
+		end
+
+		def completion
+			return false if ENV["OPS_AUTO_COMPLETE"].nil? || ENV["COMP_WORDS"].nil? || ENV["COMP_CWORD"].nil?
+
+			word_list = ENV["COMP_WORDS"].split(" ")
+			current_index = ENV["COMP_CWORD"].to_i
+			current_word = word_list[current_index]
+
+			Output.out(completion_list(current_word))
+			true
+		end
+
+		private
+
+		def bash_completion
+			"_ops_completion()  \n"\
+			"{ \n"\
+			"    COMPREPLY=( $( COMP_WORDS=\"${COMP_WORDS[*]}\" \\ \n"\
+			"                   COMP_CWORD=$COMP_CWORD \\ \n"\
+			"                   OPS_AUTO_COMPLETE=1 $1 2>/dev/null ) ) \n"\
+			"} \n"\
+			"complete -o default -F _ops_completion ops\n"
+		end
+
+		def zsh_completion
+			"function _ops_completion { \n"\
+			"	local words cword \n"\
+			"	read -Ac words \n"\
+			"	read -cn cword \n"\
+			"	reply=( $( COMP_WORDS=\"$words[*]\" \\ \n"\
+			"			   COMP_CWORD=$(( cword-1 )) \\ \n"\
+			"			   OPS_AUTO_COMPLETE=1 $words[1] 2>/dev/null )) \n"\
+			"} \n"\
+			"compctl -K _ops_completion ops\n"
+		end
+
+		def completion_list(filter)
+			(actions | builtins | forwards).select { |item| item =~ /^#{filter}/ }.sort.join(" ")
+		end
+
+		def get_forward(name, arguments)
```

Comment:
Is this being called?

Reply:
Good catch, that was for a feature I was working on where it would forward to other ops files.